docs/designs/evaluation_design.md

```diff
@@ -0,0 +1,63 @@
+# Design for ElasticDL Evaluation
+
+The document describes the ElasticDL evaluation process design.
+
+## The Existing Evaluation Design
+
+The master starts to evaluate model every many steps defined by `eval_steps` or every many seconds defined by `eval_secs`. When an evaluation job starts, the master creates evaluation tasks and inserts them into todo queue at head. Then the worker pulls an evaluation task after completing the current task. Because evaluation tasks are located at the head of todo queue.\
+The worker pulls model from parameter serve (pserver) and calculates model prediction outputs with records in the evaluation task, then it reports outputs to master. The evaluation sequence is shown as:
+
+![evaluate_sequence](/docs/images/evaluate_sequence.svg) \
+ <em>Figure 1</em>. ElasticDL Evaluate Sequence
+
+To speed up training, the worker creates `tf.data.Dataset` by [tf.data]((https://www.tensorflow.org/guide/data_performance)) API to construct an input pipeline to load data. The flowchart of train loop and evaluation in worker is shown as:
+
+![evaluate_flowchart](/docs/images/train_and_evaluate_flowchart.svg)\
+<em>Figure 2</em>. Flowchart of Train and Evaluation in Worker
+
+While the worker is performing train step N, the Dataset.prefetch thread is preparing the data for step N+1. If step N is the last batch in the task, the generator in the worker for Dataset will get the next task until the task mode is training. Then worker generates a train batch for step N+1 by Dataset.prefetch. 
+
+There are two problems in the existing evaluation design:
+
+* Problem 1: One worker will get all evaluation tasks in the flowchart shown as <em>Figure 2</em>.\
+As shown in the 10th step of <em> Figure 1.</em>, the master generates evaluation tasks and insert them into the head of todo queue when an evaluation job starts. When one worker starts to process the last mini-batch in its task, it will get next task until the task mode is training to prepare next mini-batch for Dataset.prefetch. The result is that the worker will get all evaluation tasks before it gets the next train task. Meanwhile, the master will remove all evaluation tasks from todo to doing list and other workers can not get any evaluation task. So, there is only one work to serially process evaluation tasks.
+
+* Problem 2: 
+For current ElasticDL, workers may use different model versions for evaluation for embedding vectors. For non-embedding variables, workers will use the same version from checkpoint.\
```

Comment:
workers may use different model versions for evaluation for embedding vectors => worker may use the embedding vector from different model versions for evaluation